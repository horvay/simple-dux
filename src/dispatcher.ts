export type EventType = string;
export type EventCallback = (payload: IPayload) => void;

export interface IPayload
{
    event_type: EventType;
}

export default class Dispatcher
{
    private callbacks: { [id: string]: Array<EventCallback> } = {};
    private namespace_callbacks: { [id: string]: Array<EventCallback> } = {};

    /**
      * Adds a callback to be ran for a certain event.
      */
    public addCallback(event_type: EventType, callback: EventCallback)
    {
        if (!this.callbacks[event_type])
        {
            this.callbacks[event_type] = new Array<EventCallback>();
        }

        this.callbacks[event_type].push(callback);
    }

    public addNamespaceCallback(namespace: string, callback: EventCallback)
    {
        if (!this.namespace_callbacks[namespace])
        {
            this.namespace_callbacks[namespace] = new Array<EventCallback>();
        }

        this.namespace_callbacks[namespace].push(callback);
    }

    /**
      * Sends an event into the dipatcher to be ran by the callbacks.
      */
    public injectEvent(payload: IPayload, namespace?: string)
    {
        if (namespace)
        {
            const ns_callbacks_for_event = this.namespace_callbacks[namespace];

            if (ns_callbacks_for_event)
            {
                ns_callbacks_for_event.forEach(c => c(payload));
            }
        }

        const callbacks_for_event = this.callbacks[payload.event_type];

        if (callbacks_for_event)
        {
            callbacks_for_event.forEach(c => c(payload));
        }
    }
}
