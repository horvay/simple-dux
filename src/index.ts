import Dispatcher from "./dispatcher";
import SimpleStore from "./store";

export default class SimpleDux
{
	/**
	 * The Dispatcher is used to injecct events and register callbacks for events
	 */
	public Dispatcher = new Dispatcher();

	/**
	 * The Store alllows you persist an object or register a factory to create instances
	 */
	public Store = new SimpleStore();
}
