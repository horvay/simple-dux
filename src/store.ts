type FactoryMethod = (...args) => any;

export default class SimpleStore
{
    private _persist_storage: { [key: string]: any } = {};
    private _scoped_storage: { [key: string]: FactoryMethod } = {};

    /**
     * Registers an object with the store and will return the instance whenevr retrieved
     * @param store what you want to store
     * @param name the name to reference the store
     */
    public RegisterPersistentStore<T>(store: T, name: string)
    {
        this._persist_storage[name] = store;
    }

    /**
     * Returns the instance initially stored with the given name
     * @param name the name of the object to be retrieved
     */
    public GetPersistentStore<T>(name: string): T
    {
        return this._persist_storage[name] as T;
    }

    /**
     * Registers a factory method with the storer to retrieve a new instance of an object whenever retrieved
     * @param factory_method the method that will be called to create the object
     * @param name the name given to the object to be used to retrieve it.
     */
    public RegisterScopedStore<T>(factory_method: (...args) => T, name: string)
    {
        this._scoped_storage[name] = factory_method;
    }

    /**
     * finds the factory method registered with the given name and runs it, returning the result
     * @param name the name of the object to be retrieved
     */
    public GetScopedStore<T>(name: string, ...args: any[])
    {
        let factory_method = this._scoped_storage[name];
        if (!factory_method)
        {
            throw `factory method called ${name} is not defined`;
        }

        return factory_method(...args) as T;
    }
}
