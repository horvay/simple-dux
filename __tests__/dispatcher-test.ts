import SimpleDux from "../src/index";
import { IPayload } from "../src/dispatcher";

it("dispatch an event and retrieve it", () =>
{
    let simple_dux = new SimpleDux();
    let dispatcher = simple_dux.Dispatcher;

    class TestEvent implements IPayload
    {
        public event_type = "test event";
        public test_data = "";

        constructor(data: string)
        {
            this.test_data = data;
        }
    }

    dispatcher.addCallback("test event", (event: TestEvent) =>
    {
        expect(event).toMatchSnapshot();
    });

    dispatcher.injectEvent(new TestEvent("yo momma"));
});

it("dispatch multiple events and cascade", () =>
{
    let simple_dux = new SimpleDux();
    let dispatcher = simple_dux.Dispatcher;

    class TestEvent1 implements IPayload
    {
        public event_type = "test event";
        public test_data = "";

        constructor(data: string)
        {
            this.test_data = data;
        }
    }

    class TestEvent2 implements IPayload
    {
        public event_type = "test event 2";
        public name = "no name";

        constructor(name: string)
        {
            this.name = name;
        }
    }

    dispatcher.addCallback("test event 2", (event: TestEvent2) =>
    {
        expect(event).toMatchSnapshot();
    });

    dispatcher.addCallback("test event", (event: TestEvent1) =>
    {
        expect(event).toMatchSnapshot();

        dispatcher.injectEvent(new TestEvent2("fred"));
    });

    dispatcher.injectEvent(new TestEvent1("yo momma"));
});

it("hits callback on all namespace callbacks", () =>
{
    let simple_dux = new SimpleDux();
    let dispatcher = simple_dux.Dispatcher;

    class TestEvent1 implements IPayload
    {
        public event_type = "test event";
        public test_data = "";

        constructor(data: string)
        {
            this.test_data = data;
        }
    }

    let count = 0;

    dispatcher.addNamespaceCallback("test events", (event: IPayload) =>
    {
        count++;
        expect(event).toMatchSnapshot();
    });

    dispatcher.addNamespaceCallback("test events", (event: IPayload) =>
    {
        count++;
        expect(event).toMatchSnapshot();
    });

    dispatcher.injectEvent(new TestEvent1("yo momma"), "test events");
    expect(count).toBe(2);
});