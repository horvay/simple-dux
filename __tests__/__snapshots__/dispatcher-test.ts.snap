// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dispatch an event and retrieve it 1`] = `
TestEvent {
  "event_type": "test event",
  "test_data": "yo momma",
}
`;

exports[`dispatch multiple events and cascade 1`] = `
TestEvent1 {
  "event_type": "test event",
  "test_data": "yo momma",
}
`;

exports[`dispatch multiple events and cascade 2`] = `
TestEvent2 {
  "event_type": "test event 2",
  "name": "fred",
}
`;

exports[`hits callback on all namespace callbacks 1`] = `
TestEvent1 {
  "event_type": "test event",
  "test_data": "yo momma",
}
`;

exports[`hits callback on all namespace callbacks 2`] = `
TestEvent1 {
  "event_type": "test event",
  "test_data": "yo momma",
}
`;
